axelor-base/src/main/java/com/axelor/apps/base/service/batch/BatchPasswordChange.java

```diff
@@ -0,0 +1,129 @@
+import com.axelor.apps.base.service.user.UserService;
+import com.axelor.auth.db.User;
+import com.axelor.auth.db.repo.UserRepository;
+import com.axelor.db.JPA;
+import com.axelor.db.Query;
+import com.axelor.exception.AxelorException;
+import com.axelor.exception.db.repo.ExceptionOriginRepository;
+import com.axelor.exception.service.TraceBackService;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.time.LocalDate;
+import java.util.HashMap;
+import java.util.List;
+import org.apache.commons.collections4.CollectionUtils;
+
+public class BatchPasswordChange extends AbstractBatch {
+
+  protected UserRepository userRepo;
+  protected UserService userService;
+  protected AppBaseService appBaseService;
+
+  @Inject
+  public BatchPasswordChange(
+      UserRepository userRepo, UserService userService, AppBaseService appBaseService) {
+    this.userRepo = userRepo;
+    this.userService = userService;
+    this.appBaseService = appBaseService;
+  }
+
+  @Override
+  protected void process() {
+
+    BaseBatch baseBatch = batch.getBaseBatch();
+
+    if (baseBatch.getGenerateNewRandomPasswords() || baseBatch.getUpdatePasswordNextLogin()) {
+
+      updateUsers(baseBatch);
+    }
+  }
+
+  @Transactional(rollbackOn = {AxelorException.class, Exception.class})
+  protected void updateUsers(BaseBatch baseBatch) {
+
+    LocalDate date =
+        appBaseService
+            .getTodayDate(baseBatch.getCompany())
+            .minusDays(baseBatch.getNbOfDaySinceLastUpdate());
+
+    String filter =
+        "((cast(self.passwordUpdatedOn as LocalDate) < :date OR self.passwordUpdatedOn IS NULL) AND (self.group IS NULL OR self.group.code != 'admins'))";
+
+    HashMap<String, Object> queryParameters = new HashMap<>();
+    queryParameters.put("date", date);
+
+    if (!baseBatch.getAllUsers()) {
+      filter += " AND (self.group IN (:groupSet) OR self IN (:userSet))";
+      queryParameters.put(
+          "groupSet",
+          CollectionUtils.isNotEmpty(baseBatch.getGroupSet()) ? baseBatch.getGroupSet() : 0l);
+      queryParameters.put(
+          "userSet",
+          CollectionUtils.isNotEmpty(baseBatch.getUserSet()) ? baseBatch.getUserSet() : 0l);
+    }
+
+    List<User> userList;
+    Query<User> userQuery =
+        userRepo.all().filter(filter.toString()).bind(queryParameters).order("id");
+
+    // No need of offset as user records are updating in loop
+    while (!(userList = userQuery.fetch(AbstractBatch.FETCH_LIMIT)).isEmpty()) {
+      for (User user : userList) {
+        try {
+          if (baseBatch.getGenerateNewRandomPasswords()) {
+            userService.generateRandomPasswordForUser(user);
+            user.setSendEmailUponPasswordChange(true);
+          }
+
+          if (baseBatch.getUpdatePasswordNextLogin()) {
+            user.setForcePasswordChange(true);
+          }
+          userRepo.save(user); // processChangedPassword method call in save method for email send
```

Comment:
This code can be splitted in another method. And this is this new method that should be Transactional, not updateUsers().